src/transformers/models/musicgen/modeling_musicgen.py

```diff
@@ -0,0 +1,1917 @@
+                Whether to continue running the while loop until max_length (needed for ZeRO stage 3)
+            kwargs:
+                Ad hoc parametrization of `generate_config` and/or additional model-specific kwargs that will be
+                forwarded to the `forward` function of the model. If the model is an encoder-decoder model, encoder
+                specific kwargs should not be prefixed and decoder specific kwargs should be prefixed with *decoder_*.
+
+        Return:
+            [`~utils.ModelOutput`] or `torch.LongTensor`: A [`~utils.ModelOutput`] (if `return_dict_in_generate=True`
+            or when `config.return_dict_in_generate=True`) or a `torch.FloatTensor`.
+
+                If the model is *not* an encoder-decoder model (`model.config.is_encoder_decoder=False`), the possible
+                [`~utils.ModelOutput`] types are:
+
+                    - [`~generation.GreedySearchDecoderOnlyOutput`],
+                    - [`~generation.SampleDecoderOnlyOutput`],
+                    - [`~generation.BeamSearchDecoderOnlyOutput`],
+                    - [`~generation.BeamSampleDecoderOnlyOutput`]
+
+                If the model is an encoder-decoder model (`model.config.is_encoder_decoder=True`), the possible
+                [`~utils.ModelOutput`] types are:
+
+                    - [`~generation.GreedySearchEncoderDecoderOutput`],
+                    - [`~generation.SampleEncoderDecoderOutput`],
+                    - [`~generation.BeamSearchEncoderDecoderOutput`],
+                    - [`~generation.BeamSampleEncoderDecoderOutput`]
+        """
+        if generation_config is None:
+            generation_config = self.generation_config
+
+        decoder_input_ids = kwargs.get("decoder_input_ids")
+        if decoder_input_ids is not None:
+            seq_len = decoder_input_ids.shape[-1] + 1
+        else:
+            seq_len = 1
+
+        if "max_new_tokens" in kwargs:
+            generation_config.max_new_tokens = kwargs["max_new_tokens"]
+            generation_config.max_length = kwargs["max_new_tokens"] + seq_len
+        elif "max_length" in kwargs:
+            generation_config.max_new_tokens = kwargs["max_length"] - seq_len
+            generation_config.max_length = kwargs["max_length"]
+
+        if "guidance_scale" in kwargs:
+            generation_config.guidance_scale = kwargs["guidance_scale"]
+
+        if generation_config.guidance_scale > 1:
+            cfg_processor = ClassifierFreeGuidanceLogitsProcessor(guidance_scale=generation_config.guidance_scale)
+            if logits_processor is None:
+                logits_processor = [cfg_processor]
+            else:
+                logits_processor.append(cfg_processor)
+
+        outputs = super().generate(
+            inputs,
+            generation_config,
+            logits_processor,
+            stopping_criteria,
+            prefix_allowed_tokens_fn,
+            synced_gpus,
+            **kwargs,
+        )
```

Comment:
```suggestion
        outputs = super().generate(
            inputs,
            generation_config,
            logits_processor,
            stopping_criteria,
            prefix_allowed_tokens_fn,
            synced_gpus,
            **kwargs,
        )
```
Not a fan of calling the `generate` method via `super()` here. This makes the code  difficult to reason about and creates a dependency to the "main" generate method. E.g. changing the main generate method is more complicated now as this generate method depends on it.

Generate for MusicGen really seems to be quite different, so IMO it'd be better to just fully write your own generate method here and only call:
```py
https://github.com/huggingface/transformers/blob/913899502519f3fbb7a15f52be229ff210050dc3/src/transformers/generation/utils.py#L2430
https://github.com/huggingface/transformers/blob/913899502519f3fbb7a15f52be229ff210050dc3/src/transformers/generation/utils.py#L2171
```

instead. @gante wdyt?

Reply:
Resolved in https://github.com/huggingface/transformers/pull/24109/commits/a0cae3a55b38ab4b7036f6746f5fdef3477fafa2 as per the design outlined above